machine-learning/jax-haiku-dask-dataframe-distributed-example.ipynb

```diff
@@ -0,0 +1,535 @@
+   "execution_count": null,
+   "metadata": {
+    "scrolled": true
+   },
+   "outputs": [],
+   "source": [
+    "ddf_validation[\"y_pred_no_training\"] = ddf_validation.map_partitions(dask_predict_wrapper, params=params)\n",
+    "(df_train.hvplot.scatter(x=\"x\", y=\"y\", label='Training data') * \n",
+    " ddf_validation.compute().hvplot.scatter(x=\"x\", y=\"y_pred_no_training\", label='Predicted validation data')).opts(title=\"No training\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Training - 4 examples"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "There are a number of ways we can use Dask for helping with deep learning/neural network training. For this example we implement the following which focus on when you run out of RAM/memory:\n",
+    "\n",
+    "1. Train model using Dask as a lazy loader of data.\n",
+    "2. Data-parallel training of deep learning models: Compute gradients in parallel.\n",
+    "\n",
+    "Other common use cases for distributed deep learning training include when you are CPU bound, e.g. distributed hyperparameter optimization or distributed training of an ensemble of models. This are ommited here and are left as an excercise for the reader."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### CASE 1 - Train model using Dask as a lazy loader of data"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "If we have data that is larger than the RAM we have available, we can use Dask to load and train on it batch-by-batch, one at a time.\n",
+    "\n",
+    "Let's first define an update function (the main learning function that updates the model parameters)."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "@jax.jit\n",
+    "def update(params: hk.Params, opt_state: optix.OptState, x: jnp.ndarray, y: jnp.ndarray) -> Tuple[hk.Params, optix.OptState]:\n",
+    "    grads = jax.grad(loss_function)(params, x, y)\n",
+    "    updates, opt_state = optimizer.update(grads, opt_state)\n",
+    "    params = optix.apply_updates(params, updates)\n",
+    "    return params, opt_state"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Next we define our training loop."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "scrolled": true
+   },
+   "outputs": [],
+   "source": [
+    "for epoch_number in range(200):  # num_epochs\n",
+    "    for ddf_one_partition in ddf_train.partitions:  # for each batch\n",
+    "        df_one_partition = ddf_one_partition.compute()\n",
+    "        scaled_x = jnp.array(df_one_partition[[\"scaled_x\"]].values)\n",
+    "        y = jnp.array(df_one_partition[[\"y\"]].values)\n",
+    "        params, opt_state = update(params, opt_state, scaled_x, y)"
```

Comment:
It might be worth taking a look at some of the functionality in dask-ml, which might do some of these things for you already if you're interested.

cc'ing @stsievert and @TomAugspurger